### ISSUE TITLE
validate_uniqueness_of no longer works in 2.8.0

### ISSUE THREAD

**Issue by doughsay** [thoughtbot/shoulda-matchers#679]
A test for `validate_uniqueness_of` fails with `undefined method 'validate_uniqueness_of' for #<RSpec::ExampleGroups::User:0x007ff055b73db8>`.

Reverting back to 2.7.0 causes it to work.

The relevant parts of my spec are:

```ruby
require 'rails_helper'

RSpec.describe User, :type => :model do
  let(:user) { create(:user) }
  subject { user }

  describe '#save' do
    it { is_expected.to validate_uniqueness_of(:email).case_insensitive }
  end
end
```

**Comment by mcmire** [comment]
That's strange. Just a few questions, please:

* Are you running your tests with Spring?
* Do any other matchers work?
* Have you followed the [instructions] for installing the gem?

[instructions]: https://github.com/thoughtbot/shoulda-matchers#installation

**Comment by ufchrisg** [comment]
I also ecountered this error and reverting shoulda-matchers to 2.7.0 resolved it. I am not using spring, other matchers do work (using validate_presence_of and validate_numericality_of), and my gemfile was written just as it appears in the instructions until I changed it to specify a different version for shoulda-matchers.

**Comment by doughsay** [comment]
Sorry I didn't respond earlier...

My answer to each is yes.  Although I run the tests both with and without Spring (with when developing, without so code coverage reporting works correctly).  All the other matchers I've tried work fine.  I hadn't followed the instructions exactly, but went back and changed my installation to follow it, and still no change.

Reverting to 2.7.0 is the only way I can see to restore `validate_uniqueness_of`.

**Comment by zxlai** [comment]
I also got this problem. Here is my code:

``` ruby
# spec/factories/user_spec.rb
FactoryGirl.define do
  factory :user do
    sequence(:username) { |n|  "user#{n}" }
    sequence(:email) { |n| "person#{n}@example.com" }
    password "123456"
  end
end

#spec/models/user_spec.rb
RSpec.describe User, type: :model do
  context "validations" do
    subject { FactoryGirl.build(:user) }
    it do
      should validate_presence_of(:username)
      should validate_uniqueness_of(:username).case_insensitive
    end
  end
end

#app/models/user.rb

class User < ActiveRecord::Base

  extend FriendlyId
  friendly_id :username, use: :slugged

  # Include default devise modules. Others available are:
  # :confirmable, :lockable, :timeoutable and :omniauthable
  devise :database_authenticatable, :registerable,
    :recoverable, :rememberable, :trackable, :validatable

  attr_accessor :login

  validates :username, :presence => true
  validates_uniqueness_of :username, case_insensitive: true
  validates_format_of :username, :with => /^[A-Za-z\d_]+$/, :multiline => true, :message => "can only be alphanumeric with no spaces"

  before_save { self.username = username.downcase! }

  def should_generate_new_friendly_id?
    if !slug? || username_changed?
      true
    else
      false
    end
  end

  def login=(login)
    @login = login
  end

  def login
    @login || self.username || self.email
  end

  def self.find_first_by_auth_conditions(warden_conditions)
    conditions = warden_conditions.dup
    if login = conditions.delete(:login)
      where(conditions).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
    else
      where(conditions).first
    end
  end
end

```
Here is my Guardfile
``` ruby
guard :rspec, cmd: "spring rspec" do
  require "guard/rspec/dsl"
  dsl = Guard::RSpec::Dsl.new(self)

  # Feel free to open issues for suggestions and improvements

  # RSpec files
  rspec = dsl.rspec
  watch(rspec.spec_helper) { rspec.spec_dir }
  watch(rspec.spec_support) { rspec.spec_dir }
  watch(rspec.spec_files)

  # Ruby files
  ruby = dsl.ruby
  dsl.watch_spec_files_for(ruby.lib_files)

  # Rails files
  rails = dsl.rails(view_extensions: %w(erb haml slim))
  dsl.watch_spec_files_for(rails.app_files)
  dsl.watch_spec_files_for(rails.views)

  watch(rails.controllers) do |m|
    [
      rspec.spec.("routing/#{m[1]}_routing"),
      rspec.spec.("controllers/#{m[1]}_controller"),
      rspec.spec.("acceptance/#{m[1]}")
    ]
  end

  # Rails config changes
  watch(rails.spec_helper)     { rspec.spec_dir }
  watch(rails.routes)          { "#{rspec.spec_dir}/routing" }
  watch(rails.app_controller)  { "#{rspec.spec_dir}/controllers" }

  # Capybara features specs
  watch(rails.view_dirs)     { |m| rspec.spec.("features/#{m[1]}") }

  # Turnip features and steps
  watch(%r{^spec/acceptance/(.+)\.feature$})
  watch(%r{^spec/acceptance/steps/(.+)_steps\.rb$}) do |m|
    Dir[File.join("**/#{m[1]}.feature")][0] || "spec/acceptance"
  end
end
```

Error message:
```
1) User validations should require unique value for username
     Failure/Error: should validate_uniqueness_of(:username).case_insensitive
     ActiveRecord::StatementInvalid:
       PG::NotNullViolation: ERROR:  null value in column "username" violates not-null constraint
       DETAIL:  Failing row contains (48, person1@example.com, $2a$04$N0x1xXN5ma3cIKfStEESquZ0eKo2QYS/vaOYcCFVpdI/TvYvd6tYG, null, 849af238-dac1-4d9a-96d8-f1d3de3b9e4f, null, null, null, 0, null, null, null, null, 2015-03-24 03:54:34.333744, 2015-03-24 03:54:34.333744).
       : INSERT INTO "users" ("username", "email", "encrypted_password", "slug", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"
     # ./spec/models/user_spec.rb:34:in `block (3 levels) in <top (required)>'
     # -e:1:in `<main>'
```

**Comment by zxlai** [comment]
I also got a problem when I use test for validate_uniqueness_of. Here is my code: I use `shoulda-matchers (2.8.0)`

``` ruby
# spec/factories/user_spec.rb
FactoryGirl.define do
  factory :user do
    sequence(:username) { |n|  "user#{n}" }
    sequence(:email) { |n| "person#{n}@example.com" }
    password "123456"
  end
end

#spec/models/user_spec.rb
RSpec.describe User, type: :model do
  context "validations" do
    subject { FactoryGirl.build(:user) }
    it do
      should validate_presence_of(:username)
      should validate_uniqueness_of(:username).case_insensitive
    end
  end
end

#app/models/user.rb

class User < ActiveRecord::Base

  extend FriendlyId
  friendly_id :username, use: :slugged

  # Include default devise modules. Others available are:
  # :confirmable, :lockable, :timeoutable and :omniauthable
  devise :database_authenticatable, :registerable,
    :recoverable, :rememberable, :trackable, :validatable

  attr_accessor :login

  validates :username, :presence => true
  validates_uniqueness_of :username, case_insensitive: true
  validates_format_of :username, :with => /^[A-Za-z\d_]+$/, :multiline => true, :message => "can only be alphanumeric with no spaces"

  before_save { self.username = username.downcase! }

  def should_generate_new_friendly_id?
    if !slug? || username_changed?
      true
    else
      false
    end
  end

  def login=(login)
    @login = login
  end

  def login
    @login || self.username || self.email
  end

  def self.find_first_by_auth_conditions(warden_conditions)
    conditions = warden_conditions.dup
    if login = conditions.delete(:login)
      where(conditions).where(["lower(username) = :value OR lower(email) = :value", { :value => login.downcase }]).first
    else
      where(conditions).first
    end
  end
end

```
Here is my Guardfile
``` ruby
guard :rspec, cmd: "spring rspec" do
  require "guard/rspec/dsl"
  dsl = Guard::RSpec::Dsl.new(self)

  # Feel free to open issues for suggestions and improvements

  # RSpec files
  rspec = dsl.rspec
  watch(rspec.spec_helper) { rspec.spec_dir }
  watch(rspec.spec_support) { rspec.spec_dir }
  watch(rspec.spec_files)

  # Ruby files
  ruby = dsl.ruby
  dsl.watch_spec_files_for(ruby.lib_files)

  # Rails files
  rails = dsl.rails(view_extensions: %w(erb haml slim))
  dsl.watch_spec_files_for(rails.app_files)
  dsl.watch_spec_files_for(rails.views)

  watch(rails.controllers) do |m|
    [
      rspec.spec.("routing/#{m[1]}_routing"),
      rspec.spec.("controllers/#{m[1]}_controller"),
      rspec.spec.("acceptance/#{m[1]}")
    ]
  end

  # Rails config changes
  watch(rails.spec_helper)     { rspec.spec_dir }
  watch(rails.routes)          { "#{rspec.spec_dir}/routing" }
  watch(rails.app_controller)  { "#{rspec.spec_dir}/controllers" }

  # Capybara features specs
  watch(rails.view_dirs)     { |m| rspec.spec.("features/#{m[1]}") }

  # Turnip features and steps
  watch(%r{^spec/acceptance/(.+)\.feature$})
  watch(%r{^spec/acceptance/steps/(.+)_steps\.rb$}) do |m|
    Dir[File.join("**/#{m[1]}.feature")][0] || "spec/acceptance"
  end
end
```

Error message:
```
1) User validations should require unique value for username
     Failure/Error: should validate_uniqueness_of(:username).case_insensitive
     ActiveRecord::StatementInvalid:
       PG::NotNullViolation: ERROR:  null value in column "username" violates not-null constraint
       DETAIL:  Failing row contains (48, person1@example.com, $2a$04$N0x1xXN5ma3cIKfStEESquZ0eKo2QYS/vaOYcCFVpdI/TvYvd6tYG, null, 849af238-dac1-4d9a-96d8-f1d3de3b9e4f, null, null, null, 0, null, null, null, null, 2015-03-24 03:54:34.333744, 2015-03-24 03:54:34.333744).
       : INSERT INTO "users" ("username", "email", "encrypted_password", "slug", "created_at", "updated_at") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"
     # ./spec/models/user_spec.rb:34:in `block (3 levels) in <top (required)>'
     # -e:1:in `<main>'
```

**Comment by mcmire** [comment]
@zxlai It's the `before_save` in your User model:

``` ruby
before_save { self.username = username.downcase! }
```

You need to use the non-bang version of `downcase` here; `downcase!` returns nil, so when your User is saved, it actually gets saved with no username.

**Comment by sungthecoder** [comment]
I also encountered the same error.
Answers to the questions: 
1. Yes, I use Spring, but I get the undefined error even when I turned off Spring. 
2. I didn't try all the matchers, but 'validate_presence_of' matcher works
3. Yes

Here's one interesting thing:
I have two applications using should-matchers.  Both uses 2.8.0, but in one application it works fine, but on the other, it fails with 'undefined method'.  The difference of these two applications is the one uses ActiveRecord and the other uses Mongoid
